core/vm/environment.libevm.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2024 the libevm authors.
+//
+// The libevm additions to go-ethereum are free software: you can redistribute
+// them and/or modify them under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation, either version 3 of the License,
+// or (at your option) any later version.
+//
+// The libevm additions are distributed in the hope that they will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
+// General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see
+// <http://www.gnu.org/licenses/>.
+
+package vm
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/holiman/uint256"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/libevm"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+var _ PrecompileEnvironment = (*environment)(nil)
+
+type environment struct {
+	evm           *EVM
+	self          *Contract
+	forceReadOnly bool
+}
+
+func (e *environment) ChainConfig() *params.ChainConfig  { return e.evm.chainConfig }
+func (e *environment) Rules() params.Rules               { return e.evm.chainRules }
+func (e *environment) ReadOnly() bool                    { return e.forceReadOnly || e.evm.interpreter.readOnly }
+func (e *environment) ReadOnlyState() libevm.StateReader { return e.evm.StateDB }
+func (e *environment) BlockNumber() *big.Int             { return new(big.Int).Set(e.evm.Context.BlockNumber) }
+func (e *environment) BlockTime() uint64                 { return e.evm.Context.Time }
+
+func (e *environment) Addresses() *libevm.AddressContext {
+	return &libevm.AddressContext{
+		Origin: e.evm.Origin,
+		Caller: e.self.CallerAddress,
+		Self:   e.self.Address(),
+	}
+}
+
+func (e *environment) StateDB() StateDB {
+	if e.ReadOnly() {
+		return nil
+	}
+	return e.evm.StateDB
+}
+
+func (e *environment) BlockHeader() (types.Header, error) {
+	hdr := e.evm.Context.Header
+	if hdr == nil {
+		// Although [core.NewEVMBlockContext] sets the field and is in the
+		// typical hot path (e.g. miner), there are other ways to create a
+		// [vm.BlockContext] (e.g. directly in tests) that may result in no
+		// available header.
+		return types.Header{}, fmt.Errorf("nil %T in current %T", hdr, e.evm.Context)
+	}
+	return *hdr, nil
+}
+
+func (e *environment) Call(addr common.Address, input []byte, gas uint64, value *uint256.Int, opts ...CallOption) ([]byte, uint64, error) {
+	return e.callContract(call, addr, input, gas, value, opts...)
+}
+
+func (e *environment) callContract(typ callType, addr common.Address, input []byte, gas uint64, value *uint256.Int, opts ...CallOption) ([]byte, uint64, error) {
```

Comment:
This function is the primary implementation of the PR. Everything else was a necessary refactor to fix the `Caller` and `Self` addresses for the incoming call types.